extensions-core/variance/src/main/java/io/druid/query/aggregation/variance/VarianceHolder.java

```diff
@@ -0,0 +1,227 @@
+ * <p/>
+ * where: - variance is sum[x-avg^2] (this is actually n times the variance)
+ * and is updated at every step. - n is the count of elements in chunk1 - m is
+ * the count of elements in chunk2 - t1 = sum of elements in chunk1, t2 =
+ * sum of elements in chunk2.
+ * <p/>
+ * This algorithm was proven to be numerically stable by J.L. Barlow in
+ * "Error analysis of a pairwise summation algorithm to compute sample variance"
+ * Numer. Math, 58 (1991) pp. 583--590
+ */
+public class VarianceHolder
+{
+  public static VarianceHolder from(ByteBuffer buffer)
+  {
+    return new VarianceHolder(buffer.getLong(), buffer.getDouble(), buffer.getDouble());
+  }
+
+  public static final Comparator<VarianceHolder> COMPARATOR = new Comparator<VarianceHolder>()
+  {
+    @Override
+    public int compare(VarianceHolder o1, VarianceHolder o2)
+    {
+      int compare = Longs.compare(o1.count, o2.count);
+      if (compare == 0) {
+        compare = Doubles.compare(o1.sum, o2.sum);
+        if (compare == 0) {
+          compare = Doubles.compare(o1.nvariance, o2.nvariance);
+        }
+      }
+      return compare;
+    }
+  };
+
+  static Object combineValues(Object lhs, Object rhs)
+  {
+    VarianceHolder holder1 = (VarianceHolder) lhs;
+    VarianceHolder holder2 = (VarianceHolder) rhs;
+
+    if (holder2.count == 0) {
+      return holder1;
+    }
+    if (holder1.count == 0) {
+      holder1.nvariance = holder2.nvariance;
+      holder1.count = holder2.count;
+      holder1.sum = holder2.sum;
+      return holder1;
+    }
+
+    final double ratio = holder1.count / (double) holder2.count;
+    final double t = holder1.sum / ratio - holder2.sum;
+
+    holder1.nvariance += holder2.nvariance + (ratio / (holder1.count + holder2.count) * t * t);
+    holder1.count += holder2.count;
+    holder1.sum += holder2.sum;
+
+    return holder1;
+  }
+
+  long count; // number of elements
+  double sum; // sum of elements
+  double nvariance; // sum[x-avg^2] (this is actually n times of the variance)
+
+  public VarianceHolder()
+  {
+  }
+
+  public void reset()
+  {
+    count = 0;
+    sum = 0;
+    nvariance = 0;
+  }
+
+  public VarianceHolder(long count, double sum, double nvariance)
+  {
+    this.count = count;
+    this.sum = sum;
+    this.nvariance = nvariance;
+  }
+
+  public VarianceHolder add(float v)
+  {
+    count++;
+    sum += v;
+    if (count > 1) {
+      double t = count * v - sum;
+      nvariance += (t * t) / ((double) count * (count - 1));
+    }
+    return this;
+  }
+
+  public double getVariance()
+  {
+    if (count == 0) {
+      // in SQL standard, we should return null for zero elements. But druid there should not be such a case
+      throw new IllegalStateException("should not be empty holder");
+    } else if (count == 1) {
+      return 0d;
+    } else {
+      return nvariance / count;
+    }
+  }
+
+  @JsonValue
+  public byte[] toByteArray()
+  {
+    final ByteBuffer buffer = toByteBuffer();
```

Comment:
I got the error below. This is because the length of `theBytes` is zero as `buffer.remaining()` gives zero. IMHO, we need to flip the buffer before `buffer.remaining()`.

```
java.nio.BufferUnderflowException
	at java.nio.Buffer.nextGetIndex(Buffer.java:498) ~[?:1.7.0_71]
	at java.nio.HeapByteBuffer.getLong(HeapByteBuffer.java:406) ~[?:1.7.0_71]
	at io.druid.query.aggregation.variance.VarianceHolder.from(VarianceHolder.java:53) ~[?:?]
```